java/ql/src/Telemetry/AutomodelSharedCharacteristics.qll

```diff
@@ -0,0 +1,281 @@
+float maximalConfidence() { result = 1.0 }
+
+float highConfidence() { result = 0.9 }
+
+float mediumConfidence() { result = 0.6 }
+
+/**
+ * A specification of how to  instantiate the shared characteristics for a given candidate class.
+ *
+ * The `CandidateSig` implementation specifies a type to use for Endpoints (eg., `ParameterNode`), as well as a type
+ * to label endpoint classes (the `EndpointType`). One of the endpoint classes needs to be a 'negative' class, meaning
+ *   "not any of the other known endpoint types".
+ */
+signature module CandidateSig {
+  class Endpoint;
+
+  class EndpointType;
+
+  /** Gets the string representing the file+range of the endpoint. */
+  string getLocationString(Endpoint e);
+
+  /**
+   * Defines what labels are known, and what endpoint type they correspond to.
+   */
+  predicate isKnownLabel(string label, string humanReadableLabel, EndpointType type);
+
+  /**
+   * EndpointType must have a 'negative' type that denotes the absence of any sink.
+   * This predicate should hold for that type, and that type only.
+   */
+  predicate isNegative(EndpointType t);
+
+  /**
+   * Should hold for any endpoint that is a flow sanitizer.
+   */
+  predicate isSanitizer(Endpoint e, EndpointType t);
+
+  /**
+   * Should hold for any endpoint that is a sink of the given (known or unknown) label.
+   */
+  predicate isSink(Endpoint e, string label);
+
+  /**
+   * Should hold for any endpoint that is known to not be any sink.
+   */
+  predicate isNeutral(Endpoint e);
+
+  /**
+   * Holds if `e` has the given metadata.
+   *
+   * This is a helper function to extract and export needed information about each endpoint in the sink candidate query
+   * as well as the queries that extract positive and negative examples for the prompt / training set. The metadata is
+   * extracted as a string in the format of a Python dictionary, eg.:
+   *
+   *     `{'Package': 'com.foo.util', 'Type': 'HelperClass', ... }`.
+   *
+   * The meta data will be passed on to the machine learning code by the extraction queries.
+   */
+  predicate hasMetadata(Endpoint e, string metadata);
+}
+
+/**
+ * A set of shared characteristics for a given candidate class.
+ *
+ * This module is language-agnostic, although the `CandidateSig` module will be language-specific.
+ *
+ * The language specific implementation can also further extend the behavior of this module by adding additional
+ *   implementations of endpoint characteristics exported by this module.
+ */
+module SharedCharacteristics<CandidateSig Candidate> {
+  predicate isNegative(Candidate::EndpointType e) { Candidate::isNegative(e) }
+
+  predicate isSink(Candidate::Endpoint e, string label) { Candidate::isSink(e, label) }
+
+  predicate isNeutral(Candidate::Endpoint e) { Candidate::isNeutral(e) }
+
+  /**
+   * Holds if `sink` is a known sink of type `endpointType`.
+   */
+  predicate isKnownSink(Candidate::Endpoint sink, Candidate::EndpointType endpointType) {
```

Comment:
Some of these predicates are taken from the existing extraction queries right? Could you signpost those in this PR, so I don't badger you with comments on existing code?

Reply:
The branch where I've taken these from has not been merged/reviewed — feel free to badger me, these should not be merged without unbiased review! 💪